core/src/main/java/org/testcontainers/containers/GenericContainer.java

```diff
@@ -718,21 +720,32 @@ protected Integer getLivenessCheckPort() {
 
     private void applyConfiguration(CreateContainerCmd createCommand) {
         HostConfig hostConfig = buildHostConfig();
-        createCommand.withHostConfig(hostConfig);
 
-        // Set up exposed ports (where there are no host port bindings defined)
-        ExposedPort[] portArray = exposedPorts.stream()
-                .map(ExposedPort::new)
-                .toArray(ExposedPort[]::new);
+        // PortBindings must contain:
+        //  * all exposed ports with a randomized host port (equivalent to -p CONTAINER_PORT)
+        //  * all exposed ports with a fixed host port (equivalent to -p HOST_PORT:CONTAINER_PORT)
+        List<PortBinding> allPortBindings = new ArrayList<>();
+        // First collect all the randomized host ports from our 'exposedPorts' field
+        exposedPorts.stream()
+            .map(ExposedPort::new)
+            .map(p -> new PortBinding(Ports.Binding.empty(), p))
+            .forEachOrdered(allPortBindings::add);
+        // Next collect all the fixed host ports from our 'portBindings' field
+        portBindings.stream()
+            .map(PortBinding::parse)
+            .forEachOrdered(allPortBindings::add);
```

Comment:
```suggestion
        List<PortBinding> allPortBindings = Stream
            .concat(
                // First collect all the randomized host ports from our 'exposedPorts' field
                exposedPorts.stream().map(Object::toString),

                // // Next collect all the fixed host ports from our 'portBindings' field
                portBindings.stream()
            )
            .map(PortBinding::parse)
            .collect(Collectors.toList());
```

Reply:
Probabilistic bug found while trying this out; replaced by a Map-based approach which ensures uniqueness of bound ports.